src/Microsoft.AspNetCore.Routing/DefaultEndpointRouteBuilder.cs

```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Builder;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class DefaultEndpointRouteBuilder : IEndpointRouteBuilder
+    {
+        public DefaultEndpointRouteBuilder()
+        {
+            EndpointDataSources = new List<EndpointDataSource>();
+        }
+
+        public IApplicationBuilder ApplicationBuilder { get; set; }
+
+        public IApplicationBuilder CreateApplicationBuilder() => ApplicationBuilder.New();
```

Comment:
@davidfowl re earlier question about this: So you can easily turn existing middleware into an endpoint

```cs
public static class EndpointDataSourceBuilderExtensions
{
    public static IApplyEndpointBuilder MapHello(this EndpointDataSourcesBuilder builder, string template, string greeter)
    {
        if (builder == null)
        {
            throw new ArgumentNullException(nameof(builder));
        }

        var pipeline = builder.CreateApplicationBuilder()
           .UseHello(greeter)
           .Build();

        return builder.Map(
            template,
            "Hello " + greeter,
            pipeline);
    }
}
```

Good? Bad? Better way of doing it?

Reply:
> Good? ~~Bad?~~ Better way of doing it?

This is elegant and cool. Just like `Map` for middleware but it uses routing. Any middleware or middware pipeline can be 'endpointed'